ta/trusted_keys/entry.c

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019-2020, Linaro Limited
+ */
+
+#include <assert.h>
+#include <pta_system.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+#include <trusted_keys.h>
+#include <util.h>
+
+#define IV_SIZE			16
+#define TAG_SIZE		16
+#define MAX_BUF_SIZE		512
+
+/*
+ * Acronym:
+ *
+ * TK - Trusted Key
+ */
+
+struct tk_blob_hdr {
+	uint8_t reserved;
+	uint8_t iv[IV_SIZE];
+	uint8_t tag[TAG_SIZE];
+	uint8_t enc_key[];
+};
+
+static TEE_Result get_random(uint32_t types, TEE_Param params[TEE_NUM_PARAMS])
+{
+	DMSG("Invoked TA_CMD_GET_RANDOM");
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!params[0].memref.buffer || !params[0].memref.size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	TEE_GenerateRandom(params[0].memref.buffer, params[0].memref.size);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result derive_unique_key(uint8_t *key, uint16_t key_size,
+				    uint8_t *extra, uint16_t extra_size)
+{
+	TEE_TASessionHandle sess = TEE_HANDLE_NULL;
+	TEE_Param params[TEE_NUM_PARAMS] = { };
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t ret_orig = 0;
+	uint32_t param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+					       TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					       TEE_PARAM_TYPE_NONE,
+					       TEE_PARAM_TYPE_NONE);
+
+	res = TEE_OpenTASession(&(const TEE_UUID)PTA_SYSTEM_UUID,
+				TEE_TIMEOUT_INFINITE, 0, NULL, &sess,
+				&ret_orig);
+	if (res)
+		return res;
+
+	if (extra && extra_size) {
+		params[0].memref.buffer = extra;
+		params[0].memref.size = extra_size;
+	}
+
+	params[1].memref.buffer = key;
+	params[1].memref.size = key_size;
+
+	res = TEE_InvokeTACommand(sess, TEE_TIMEOUT_INFINITE,
+				  PTA_SYSTEM_DERIVE_TA_UNIQUE_KEY,
+				  param_types, params, &ret_orig);
+
+	TEE_CloseTASession(sess);
+
+	return res;
+}
+
+static TEE_Result huk_ae_encrypt(TEE_OperationHandle crypto_op, uint8_t *in,
+				 uint32_t in_sz, uint8_t *out, uint32_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct tk_blob_hdr *hdr = (struct tk_blob_hdr *)out;
+	uint32_t enc_key_len = in_sz;
+	uint32_t tag_len = TAG_SIZE;
+
+	hdr->reserved = 0;
+	TEE_GenerateRandom(hdr->iv, IV_SIZE);
+
+	res = TEE_AEInit(crypto_op, hdr->iv, IV_SIZE, TAG_SIZE * 8, 0, 0);
+	if (res)
+		return res;
+
+	res = TEE_AEEncryptFinal(crypto_op, in, in_sz, hdr->enc_key,
+				 &enc_key_len, hdr->tag, &tag_len);
+	if (res || tag_len != TAG_SIZE)
+		return TEE_ERROR_SECURITY;
+
+	if (ADD_OVERFLOW(enc_key_len, sizeof(*hdr), out_sz))
```

Comment:
sorry: late comment, and maybe to discard.
Yet. I wonder if this should be tested before the calls above write to the output buffer.

Reply:
Prior test has already been done as part of `seal/unseal_trusted_key()` APIs. This is just to check if `enc_key_len` returned from `TEE_AEEncryptFinal()` doesn't lead to any overflow.

Reply:
ok. fine.